crates/llm-local/src/token_output_stream.rs

```diff
@@ -0,0 +1,68 @@
+/// Implementation for TokenOutputStream Code is borrowed from
+/// https://github.com/huggingface/candle/blob/main/candle-examples/src/token_output_stream.rs
+/// (Commit SHA 4fd00b890036ef67391a9cc03f896247d0a75711)
+///
+/// This is a wrapper around a tokenizer to ensure that tokens can be returned to the user in a
+/// streaming way rather than having to wait for the full decoding.
+pub struct TokenOutputStream {
+    tokenizer: tokenizers::Tokenizer,
+    tokens: Vec<u32>,
+    prev_index: usize,
+    current_index: usize,
+}
+
+impl TokenOutputStream {
+    pub fn new(tokenizer: tokenizers::Tokenizer) -> Self {
+        Self {
+            tokenizer,
+            tokens: Vec::new(),
+            prev_index: 0,
+            current_index: 0,
+        }
+    }
+
+    fn decode(&self, tokens: &[u32]) -> anyhow::Result<String> {
+        match self.tokenizer.decode(tokens, true) {
+            Ok(str) => Ok(str),
+            Err(err) => anyhow::bail!("cannot decode: {err}"),
+        }
+    }
+
+    /// Processes the next token in the sequence, decodes the current token stream,
+    /// and returns any newly decoded text.
+    /// https://github.com/huggingface/text-generation-inference/blob/5ba53d44a18983a4de32d122f4cb46f4a17d9ef6/server/text_generation_server/models/model.py#L68
+    pub fn next_token(&mut self, token: u32) -> anyhow::Result<Option<String>> {
+        let prev_text = if self.tokens.is_empty() {
+            String::new()
+        } else {
+            let tokens = &self.tokens[self.prev_index..self.current_index];
+            self.decode(tokens)?
+        };
+        self.tokens.push(token);
+        let text = self.decode(&self.tokens[self.prev_index..])?;
+        if text.len() > prev_text.len() && text.chars().last().unwrap().is_alphanumeric() {
```

Comment:
I don't fully understand what this check is supposed to be doing. Why do we care about the length of the next text vs the previous, and why do we care whether the last character is alphanumeric?

Reply:
The length check is to see if we have any new tokens. The alphanumeric check is supposed to be to check if we have a valid token to decode. That is what I gather from the python function the docs link to 

https://github.com/huggingface/text-generation-inference/blob/5ba53d44a18983a4de32d122f4cb46f4a17d9ef6/server/text_generation_server/models/model.py#L68

Reply:
The python code is dealing with unfinished utf-8 byte sequences which is not possible at this point in the Rust code. Rust `char`s are guaranteed to be valid utf-8. The check for alphanumeric chars is checking that the character is `A-Z | a-z | 0-9` which does seem to be what we want.

The `Tokenizer::decode` function returns `String`s so I'm guessing somehow the `tokenizer` crate is taking care of byte sequences that aren't valid utf-8?

Reply:
Here is the relevant rust version from where this is borrowed.
https://github.com/huggingface/candle/blob/6eea45a761fc1636b5e8012d02bdaa93321652ca/candle-examples/src/token_output_stream.rs#L43